src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerations/AvoidMultipleEnumerations.cs

```diff
@@ -0,0 +1,256 @@
+            "All",
+            "Any",
+            "Average",
+            "Contains",
+            "Count",
+            "DefaultIfEmpty",
+            "ElementAt",
+            "ElementAtOrDefault",
+            "First",
+            "FirstOrDefault",
+            "Last",
+            "LastOrDefault",
+            "LongCount",
+            "Max",
+            "Min",
+            "Single",
+            "SingleOrDefault",
+            "Sum",
+            "ToArray",
+            "ToDictionary",
+            "ToHashSet",
+            "ToList",
+            "ToLookup");
+
+        /// <summary>
+        /// Linq methods causing its first two parameters to be enumerated.
+        /// </summary>
+        private static readonly ImmutableArray<string> s_linqTwoParametersEnumeratedMethods = ImmutableArray.Create(
+            "SequenceEqual");
+
+        /// <summary>
+        /// Linq methods deferring its first parameter to be enumerated.
+        /// </summary>
+        private static readonly ImmutableArray<string> s_linqOneParameterDeferredMethods = ImmutableArray.Create(
+            "Append",
+            "Cast",
+            "Distinct",
+            "GroupBy",
+            "OfType",
+            "OrderBy",
+            "OrderByDescending",
+            "Prepend",
+            "Reverse",
+            "Select",
+            "SelectMany",
+            "Skip",
+            "SkipLast",
+            "SkipWhile",
+            "Take",
+            "TakeLast",
+            "TakeWhile",
+            "ThenBy",
+            "ThenByDescending",
+            "Where");
+
+        /// <summary>
+        /// Linq methods deferring its first two parameters to be enumerated.
+        /// </summary>
+        private static readonly ImmutableArray<string> s_linqTwoParametersDeferredMethods = ImmutableArray.Create(
+            "Concat",
+            "Except",
+            "GroupJoin",
+            "Intersect",
+            "Join",
+            "Union");
+
+        internal abstract GlobalFlowStateDictionaryFlowOperationVisitor CreateOperationVisitor(
+            GlobalFlowStateDictionaryAnalysisContext context,
+            WellKnownSymbolsInfo wellKnownSymbolsInfo,
+            IMethodSymbol? getEnumeratorMethod);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.RegisterCompilationStartAction(CompilationStartAction);
+        }
+
+        private void CompilationStartAction(CompilationStartAnalysisContext context)
+        {
+            context.RegisterOperationBlockStartAction(OnOperationBlockStart);
+        }
+
+        private void OnOperationBlockStart(OperationBlockStartAnalysisContext operationBlockStartAnalysisContext)
+        {
+            var wellKnownTypeProvider = WellKnownTypeProvider.GetOrCreate(operationBlockStartAnalysisContext.Compilation);
```

Comment:
This is hitting a cache lookup, we should avoid doing this inside each operation block start. Instead get the provider in `CompilationStartAction` and pass it as a parameter to `OnOperationBlockStart`

Reply:
Done